### ISSUE TITLE
Cannot compile x86 GRPC 1.8 or 1.10 on Windows 10 Platform

### ISSUE THREAD

**Issue by mastergberry** [grpc/grpc#14726]
Hi. I am having issues compiling the new GRPC platform on my local machine. I am going to tag @jtattermusch since he helped me last time and was the one that setup all the Windows CMake stuff a year ago when it was not very well documented yet.

I am using Visual Studios Enterprise 2015. Version 14.0.25431.01 Update 3

I followed the instructions on both the 1.8.x and 1.10.x branches and have been getting this same pop up each time:

![image](https://user-images.githubusercontent.com/1483171/37516938-2a2e8dba-2910-11e8-9e7b-35622c12a90d.png)

It freezes on this section of the compilation steps:

```
-- Found Threads: TRUE
-- Found Perl: C:/Perl64/bin/perl.exe (found version "5.24.3")
-- Check size of __int32
-- Check size of __int32 - done
-- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.10.1.windows.1")
-- git Version: v1.3.5-474c5950
-- Version: 1.3.5
-- Performing Test HAVE_STD_REGEX
```

As soon as I click "Ok" it spits out this information

```
-- Performing Test HAVE_STD_REGEX -- compiled but failed to run
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- failed to compile
CMake Error at third_party/benchmark/CMakeLists.txt:183 (message):
  Failed to determine the source files for the regular expression backend


-- Configuring incomplete, errors occurred!
```

I have attempted to use the old cmake build method with Ninja and the new Visual Studios solution file build instructions. I even upgraded Go/Perl/Ninja because I thought there might have been something new i needed in a more recent version.

I would like to note that I was able to compile x64 without any issues when using Ninja. I only have been having issues when trying to build x86 or trying to build for both (the Visual Studios approach).

Below are two gists for my cmake logs which are labeled.

CMakeOutput.log: https://gist.github.com/mastergberry/464b1630acf9694a34a8f1cdc3e82e4d
CMakeError.log: https://gist.github.com/mastergberry/6f15b6be7211c3d658381b3563494394

Happy to provide any other information or logs that might be useful. Thanks.

**Comment by jtattermusch** [comment]
@mastergberry   does the x86 build work well with VS only build (= not using Ninja)?   We have a continuous test for x86 build on Visual Studio (which is passing consistently), but we are not testing testing with Ninja directly.

**Comment by mastergberry** [comment]
@jtattermusch Hey, if you mean these steps from the `INSTALL.md` file:

```
> @rem Run from grpc directory after cloning the repo with --recursive or updating submodules.
> md .build
> cd .build
> cmake .. -G "Visual Studio 14 2015" -DCMAKE_BUILD_TYPE=Release
> cmake --build .
```

Then yes I tried using this too and it did not work for me. It was crashing on the same spot as the ninja build was.

**Comment by mastergberry** [comment]
Here are the logs from using the Visual Studio build approach:

CMakeError.log: https://gist.github.com/mastergberry/698c426b005a23f57e595c62a1cf7104
CMakeOutput.log: https://gist.github.com/mastergberry/2da412bcc0a474a6124c44cd6ef9b8dc

**Comment by jtattermusch** [comment]
@mastergberry   from what you're saying it looks more like the "Performing Test HAVE_STD_REGEX" fails for some reason - which would point to a cmake issue. Which cmake version are you using?

**Comment by mastergberry** [comment]
@jtattermusch 

```
cmake version 3.7.2

CMake suite maintained and supported by Kitware (kitware.com/cmake)
```

Is there a recommended version of cmake? I would have imagined that the cmake files would have complained earlier on about outdated versions (I thought it checks for this in the Makefiles)

**Comment by jtattermusch** [comment]
I also have cmake 3.7.2 (anything >2.8.x should work I believe)

**Comment by jtattermusch** [comment]
@mastergberry   if you could do git bisect to see which commit introduced this problem, it would be great.

**Comment by mastergberry** [comment]
@jtattermusch I'm not familiar with this command. I looked it up briefly and it looks like I need to

1) type `git bisect start` on the current version
2) type `git bisect bad` on the current version
3) type `git bisect good [hash of working build]`

and then it will do some kind of binary search between the number of commits to find the bad commit by checking out random commits and me building each one right?

**Comment by jtattermusch** [comment]
@mastergberry   yes, it allows you to find out which commit introduced a suspected breakage (=> you're saying that the build worked for you earlier and then got broken). in a reasonable (logarithmic) number of steps.

**Comment by mastergberry** [comment]
@jtattermusch  After all of my testing I realized later on that this command `call "%VS140COMNTOOLS%..\..\VC\vcvarsall.bat" x86` was giving the following output instead of nothing

```
The input line is too long.
The syntax of the command is incorrect.
```

The last command i typed was `git bisect good` and I got this output

```
commit 438c0b5771d748df299a75a844fa70c8e8539b5a
Author: Jan Tattermusch <jtattermusch@google.com>
Date:   Mon Jan 23 11:24:53 2017 +0100

    add convenience cmake targets

:100644 100644 1b66ee7894c60cd09725790fa5e68ff3d218059c 41d677d4d24827b88ce5635096ba1c4b1fa93b9c M  CMakeLists.txt
:040000 040000 17a453fd172cc9d19f59c0a6e66d6b355f667937 e2aee457f5b5121f1e9e59f93f5017adb2819f3e M  templates
```

**Comment by mastergberry** [comment]
Just noting it was not always saying 

```
The input line is too long.
The syntax of the command is incorrect.
```

Just on some of the older builds when i got far enough in the `git bisect`

**Comment by jtattermusch** [comment]
@mastergberry   I think the `The input line is too long.` error is unrelated - it happens when you the vcvarsall.bat is called multiple times and keeps appending to the env variables until they grow too large.

**Comment by mastergberry** [comment]
@jtattermusch ok do u need anything else from me atm?

**Comment by Kalmalyzer** [comment]
I do not have an explanation, but this should at least narrow it down:

grpc includes the [google/benchmark](https://github.com/google/benchmark) library. This has a couple of regex tests. It is one of those regex tests that is failing for @mastergberry .

Here is the console output when the configuration succeeds with VS2015:
```
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- failed to compile
```

The error message that @mastergberry provides a screenshot of, and the message `-- Performing Test HAVE_STD_REGEX -- compiled but failed to run` indicates that CMake compiles a small test program, but the OS then fails to run it.

The failing feature detection step is invoked from [this CMakeLists.txt file](https://github.com/google/benchmark/blob/e668e2a1baaa969a8e5a9220e3949e3202dea9e8/CMakeLists.txt#L187).
It has a couple of lines which [attempt to compile and run a small test program](https://github.com/google/benchmark/blob/e668e2a1baaa969a8e5a9220e3949e3202dea9e8/cmake/CXXFeatureCheck.cmake#L45-L48).

The [source file](https://github.com/google/benchmark/blob/e668e2a1baaa969a8e5a9220e3949e3202dea9e8/cmake/std_regex.cpp) for the test program is quite short.

Then [this line](https://github.com/google/benchmark/blob/e668e2a1baaa969a8e5a9220e3949e3202dea9e8/cmake/CXXFeatureCheck.cmake#L59) prints an error under the unusual condition that the test program compiled, but the OS was unable to start the application. This is normally only something that [people run into when they cross-compile](https://github.com/google/benchmark/issues/351), because then they will compile for a different OS/CPU architecture than they are currently on -- and the compiled result will naturally not be executable under their host OS.

So... It appears that the small example program that is being built by CMake is not startable on @mastergberry 's machine. There are some tips on [how to debug try_compile & try_run problems](https://cmake.org/pipermail/cmake/2011-August/046044.html).

It so happens that for google/benchmark, launching cmake multiple times will re-run the regex dection steps (and only those steps). Therefore, when I cmake with this command line: `cmake .. -G "Visual Studio 14 2015" -DCMAKE_BUILD_TYPE=Release --debug-trycompile` , I get just the relevant files in `Release\CMakeFiles\CMakeTmp`. A bit of free-text searching allows me to pinpoint which exact files/folders are relevant to this particular problem.

@mastergberry If you do a debug run, you will be able to find Visual Studio solution files, log files from the compilation/link steps, and the final test executables, all in that directory. If you inspect those, perhaps you can find out why one of the test executables is not possible to launch on your machine.

**Comment by mastergberry** [comment]
So...I reformatted my machine (it was a mess...it needed to be done) and now I am having no issues getting the first cmake command to run. When I try to run the second cmake command though I am getting some errors related to `vsyasm.exe` returning 1 on boringssl stuff.

I remember I had a similar issue a while back...this thing was weird to install. I tried copying it into the locations from the readme but doesn't seem to help. Any stupid/easy thing I am doing wrong?
